ResearchKit/Common/ORKOrderedTask.m

```diff
@@ -308,8 +308,21 @@ + (ORKCompletionStep *)makeCompletionStep {
     return step;
 }
 
+static void ORKValidateStepUniquiness(NSMutableArray *array, ORKStep *step) {
+    NSPredicate *uniqueIdPredicate = [NSPredicate predicateWithBlock:^BOOL(ORKStep *evaluatedStep, NSDictionary *bindings) {
+        return [evaluatedStep.identifier isEqualToString:step.identifier];
+    }];
+    
+    NSArray *duplicateSteps = [array filteredArrayUsingPredicate:uniqueIdPredicate];
+    
+    if (duplicateSteps.count) {
+        @throw [NSException exceptionWithName:NSGenericException reason:@"All steps should have unique identifier" userInfo:nil];
+    }
+}
+
 static void ORKStepArrayAddStep(NSMutableArray *array, ORKStep *step){
     [step validateParameters];
+    ORKValidateStepUniquiness(array, step);
```

Comment:
I don't think I'd do this here. Instead, I'd do it by having adding an optional method to `ORKTask` protocol, implemented by `ORKOrderedTask`, called `-validateParameters`. This would check for unique identifiers. You'd call it in the `task` setter on `ORKTaskViewController`, which would match the `ORKStepViewController` implementation.